### PR TITLE
tpr storage int tests

### LINKED ISSUE #407: Add TPR storage to the tests in test/integration/clientset_test.go
They currently test just the etcd-based server. I believe there is a bug with service class creation as well, so this will force fixing that bug.

Left to do:

- [x] Mock out the client interface and pass it to the TPR options in integration.go
    - Probably using `github.com/kubernetes-incubator/service-catalog/pkg/client/clientset_generated/clientset/fake` -- requires #505, as @krancour is writing a mock `RESTClient` that this fake will need to use.
- [ ] Re-enable the TPR storage type in clientset_test.go

### LINKED ISSUE #581: Enable controller tests for TPR storage

### PR DESCRIPTION
**Pull request by krancour**
Closes #407 
Closes #581 

#599 is a critical follow-up to this.

__Edit:__ Since it's been requested, here's a brief overview of what this PR adds... all it does is enables an _existing_ suite of storage tests (which are already run against etcd-based storage) to _also_ be run against tpr-based storage. The exception to this is that test suite terminates early when tpr-based storage is being tested in order to avoid failing due to known issues. (The idea here is that some coverage of the tpr-based storage is better than none while we continue to sort out the remaining issues.) To facilitate these tests, which have a dependency on the upstream apiserver (to the extent it is used to store the tpr analogs to our own types), this test introduces a mocked out apiserver _restclient_. Instead of talking to an actual apiserver, that restclient short-circuits all requests by muxing them in-process and dispatching to handlers that decode/store and retrieve/encode as appropriate.

### CODE REVIEW

File: test/integration/clientset_test.go [hunk 1]
```diff
@@ -219,6 +218,14 @@ func testBrokerClient(client servicecatalogclient.Interface, name string) error
 		)
 	}
 
+	// TODO: Here be dragons. Tests fail beyond this point due to known issues
+	// with our TPR-based storage implementation. Bail early (until those issues)
+	// are fixed, because some tests are better than no tests. If storage isn't
+	// TPR-based, carry on.
```
Comment (arschles) [inline test/integration/clientset_test.go]: for those watching, this work is captured in https://github.com/kubernetes-incubator/service-catalog/issues/599. @krancour is in-progress on that issue

File: test/integration/clientset_test.go [hunk 2]
```diff
@@ -390,6 +397,14 @@ func testServiceClassClient(client servicecatalogclient.Interface, name string)
...
```
Comment (arschles) [inline test/integration/clientset_test.go]: same here RE #599, and below

File: test/integration/fake_core_rest_client.go [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+	if err != nil {
+		log.Fatalf("error encoding list: %s", err)
+	}
+	rw.Write(listBytes)
+}
+
+func createItem(rw http.ResponseWriter, r *http.Request) {
+	ns := mux.Vars(r)["namespace"]
+	tipe := mux.Vars(r)["type"]
+	// TODO: Is there some type-agnostic way to get the codec?
```
Comment (arschles) [inline test/integration/fake_core_rest_client.go]: @pmorie do you know the answer to this question?

File: test/integration/clientset_test.go (hunk 1 above)
Comment (MHBauer) [inline test/integration/clientset_test.go]: this seems like a reason to pass the testing.T in so we can explicitly call `t.Skip("reason")` along with all of the other error and logging functions.

**Comment by krancour** [comment]
@MHBauer 

> why yet another fake?

Because upstream doesn't seem to supply a fake restclient and the tpr-based storage relies, specifically, on a restclient.

**Comment by pmorie** [comment]
Trying to get jenkins to test this as

**Comment by vaikas-google** [comment]
I like @MHBauer suggestion of using the Skip, is that doable, or is there a reason to not do that? Other than that LGTM.

**Comment by pmorie** [comment]
We can use `Skip` in a follow up - functionally this looks like it pushes us forward

**Pull request merged**